mithril-common/src/crypto_helper/merkle_map.rs

```diff
@@ -0,0 +1,617 @@
+    }
+}
+
+impl<K: MKMapKey> From<MKProof> for MKMapProof<K> {
+    fn from(other: MKProof) -> Self {
+        MKMapProof::new(other, BTreeMap::default())
+    }
+}
+
+/// A MKMap node
+#[derive(Clone)]
+pub enum MKMapNode<K: MKMapKey> {
+    /// A MKMap node
+    Map(Rc<MKMap<K, Self>>),
+
+    /// A MKTree node
+    Tree(Rc<MKTree>),
+
+    /// A MKTreeNode node
+    TreeNode(MKTreeNode),
+}
+
+impl<K: MKMapKey> MKMapValue<K> for MKMapNode<K> {
+    fn compute_root(&self) -> StdResult<MKTreeNode> {
+        match self {
+            MKMapNode::Map(mk_map) => mk_map.compute_root(),
+            MKMapNode::Tree(merkle_tree) => merkle_tree.compute_root(),
+            MKMapNode::TreeNode(merkle_tree_node) => Ok(merkle_tree_node.to_owned()),
+        }
+    }
+
+    fn contains<T: Into<MKTreeNode> + Clone>(&self, leaf: &T) -> bool {
+        let leaf = leaf.to_owned().into();
+        match self {
+            MKMapNode::Map(mk_map) => mk_map.contains(&leaf).is_some(),
+            MKMapNode::Tree(merkle_tree) => merkle_tree.contains(&leaf),
+            MKMapNode::TreeNode(merkle_tree_node) => *merkle_tree_node == leaf,
+        }
+    }
+
+    fn compute_proof<T: Into<MKTreeNode> + Clone>(
+        &self,
+        leaves: &[T],
+    ) -> StdResult<Option<MKMapProof<K>>> {
+        match self {
+            MKMapNode::Tree(ref value) => {
+                let proof = value
+                    .compute_proof(
+                        &leaves
+                            .iter()
+                            .map(|leaf| leaf.to_owned().into())
+                            .collect::<Vec<_>>(),
+                    )
+                    .with_context(|| "MKMapValue could not compute sub proof for MKTree")?;
+                Ok(Some(proof.into()))
+            }
+            MKMapNode::Map(ref value) => {
+                let proof = value
+                    .compute_proof(
+                        &leaves
+                            .iter()
+                            .map(|leaf| leaf.to_owned().into())
+                            .collect::<Vec<_>>(),
+                    )
+                    .with_context(|| "MKMapValue could not compute sub proof for MKMap")?;
+                Ok(Some(proof))
+            }
+            _ => Ok(None),
+        }
+    }
+}
+
+impl<K: MKMapKey> From<MKMap<K, MKMapNode<K>>> for MKMapNode<K> {
+    fn from(other: MKMap<K, MKMapNode<K>>) -> Self {
+        MKMapNode::Map(Rc::new(other))
+    }
+}
+
+impl<K: MKMapKey> From<MKTree> for MKMapNode<K> {
+    fn from(other: MKTree) -> Self {
+        MKMapNode::Tree(Rc::new(other))
+    }
+}
+
+impl<K: MKMapKey> From<MKTreeNode> for MKMapNode<K> {
+    fn from(other: MKTreeNode) -> Self {
+        MKMapNode::TreeNode(other)
+    }
+}
+
+impl<K: MKMapKey> TryFrom<MKMapNode<K>> for MKTreeNode {
+    type Error = StdError;
+    fn try_from(other: MKMapNode<K>) -> Result<Self, Self::Error> {
+        other.compute_root()
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
General comments for the tests:
Why do you need to generate `MKMap` with 10 of thousands of values for unit tests ? It would be more understandable to use by hands maps (additionally it would allow to use have console logs of assert_eqs that explain exactly what's wrong).

Reply:
I reduced the number of entries in the tests :+1: